tcmur_cmd_handler.c

```diff
@@ -628,26 +584,58 @@ static int handle_passthrough(struct tcmu_device *dev,
 	return ret;
 }
 
-/*
- * try to passthrough the command if handler supports command passthrough.
- * note that TCMU_NOT_HANDLED is returned when a tcmur handler does not
- * handle a passthrough command, but since we call ->handle_cmd via
- * async_call_command(), ->handle_cmd can finish in the callers context
- * (asynchronous handler) or work queue context (synchronous handlers),
- * thus we'd need to check if ->handle_cmd handled the passthough command
- * here as well as in handle_passthrough_cbk().
- */
+bool tcmur_handler_is_passthrough_only(struct tcmur_handler *rhandler)
+{
+	if (rhandler->write || rhandler->read || rhandler->flush)
+		return false;
+
+	return true;
+}
+
 int tcmur_cmd_handler(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 {
-	int ret;
+	int ret = TCMU_NOT_HANDLED;
 	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
 	struct tcmur_handler *rhandler = handler->hm_private;
+	uint8_t *cdb = cmd->cdb;
+	struct iovec *iovec = cmd->iovec;
+	size_t iov_cnt = cmd->iov_cnt;
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	off_t offset = block_size * tcmu_get_lba(cdb);
 
-	if (rhandler->handle_cmd) {
-		ret = handle_passthrough(dev, rhandler, cmd);
-		if ((ret == TCMU_ASYNC_HANDLED) || (ret != TCMU_NOT_HANDLED))
-			return ret;
```

Comment:
Ah, I think I know what you mean for some of this.

For handlers that do not do all passthrough like file_optical, handle_passthrough is now called at the bottom of the command handling switch instead of before it. So instead of allowing handlers to override all commands, it now only allows handlers to implement commands that are not handled by the generic tcmu-runner SCSI code. Is that what you mean by your first comment about dropping the command passthrough out? If so, that was a mistake. We can add the original behavior back.

From there I am guessing you are saying I should do RTPG/STPG handling in rbd.. It sounds like you think I am making it ceph specific because I added callouts which only rbd.c implemented? For this, I disagree. Besides other handlers working on support and ALUA being generic for all other handlers, we can either have handlers:

1. Call the alua.c code from their handle_cmd callout.
2. Have tcmu-runner call the alua code and have runner call into the handlers.

I do not see an advantage to #1, but am open to a discussion.

I preferred #2 because we only have to have one caller of tcmu_emulate_set_tgt_port_grps() that we need to worry about. If we have tcmu-runner calling it and individual handlers, it makes it more of a pain to update later like the write verify code. I did make tcmu_emulate_set_tgt_port_grps a libtcmu function to allow outside daemons to use it since it seemed like it could be useful, but we do not have to worry about them.

It is like how you only converted a couple handlers to the read/write/flush callouts and then we had a mess of different behaviors. That makes maintaining multiple handlers difficult for me.

I also prefer tcmu-runner to do the ALUA calls because it needs to be coordinated with other commands like inquiry. I do not want handlers to be making alua.c calls then having to figure out how to handle port id changes. tcmu-runner should coordinate that. The handler just needs to handle the low level details.